src/build/env.py

```diff
@@ -177,35 +266,56 @@ def install(self, requirements: Collection[str]) -> None:
                 os.path.abspath(req_file.name),
             ]
             run_subprocess(cmd)
+
         finally:
             os.unlink(req_file.name)
 
+    @property
+    def name(self) -> _EnvImplInclDefault:
+        return 'virtualenv+pip' if self._create_with_virtualenv else 'venv+pip'
 
-def _create_isolated_env_virtualenv(path: str) -> tuple[str, str]:
-    """
-    We optionally can use the virtualenv package to provision a virtual environment.
 
-    :param path: The path where to create the isolated build environment
-    :return: The Python executable and script folder
-    """
-    import virtualenv
+class _UvImplBackend(_EnvImplBackend):
+    def create(self, path: str) -> None:
+        import venv
 
-    if _valid_global_pip():
-        cmd = [path, '--no-seed', '--activators', '']
-    else:
-        cmd = [path, '--no-setuptools', '--no-wheel', '--activators', '']
+        self._env_path = path
+
+        # ``uv.find_uv_bin`` will look for uv in the user prefix if it can't
+        # find it under ``sys.prefix``, essentially potentially rearranging
+        # the user's $PATH.  We'll only look for uv under the prefix of
+        # the running interpreter for unactivated venvs then defer to $PATH.
+        uv_bin = shutil.which('uv', path=sysconfig.get_path('scripts'))
+
+        if not uv_bin:
+            uv_bin = shutil.which('uv')
```

Comment:
I strongly disagree with such fallback. Either pick uv from sys.prefix or fail.

Reply:
I've generally been respecting the design of uv, which is to _not_ install itself in environments, but to be used globally (via https://astral.sh/uv/install.sh, cargo install, homebrew, pacman, etc) and only target environments. If a user does install uv in build's environment, that should take preference, but I think we should try to use an outer/system `uv` otherwise. I don't think someone passing `--installer=uv` would be unhappy if a `uv` executable on their path is used. (same with `cmake`, `gcc`, `ninja`, `patchelf`, etc., lots of executables get used from the path)

But willing to go with this if you really prefer.

Reply:
Then remove the UV extra because user should install it separately? Having an extra but then taking a global is unexpected to me.

Reply:
A user _can_ install it, say if the system `uv` is older than they like, or if they just use `pipx run build[uv]` regardless of if the system has uv or not. But, if they are on a system that does not support binary wheels, like BSDs, they might prefer or even need to install it globally and not use the extra.

(This is very common for binary deps with optional PyPI packages like CMake and ninja)

Reply:
We are not in the business of installing cmake, why is UV any different (other than it happens to be available on pypi, is the same a random binary).

Reply:
I’m saying it is similar (any binary CLI tool with a PyPA distribution), and this what experience has shown is the most useful way to handle it. PyPI package preferred, binary on path second is used in scikit-build and many hand written packages. Otherwise you can’t control the CMake version on binary supported platforms or you can’t build at all on non binary supported platforms.

Also, other distributions of uv might not include this Python package, the CMake Python wrapper is not included in conda-forge, for example.

Reply:
I don't see why we have to support that use case. From my point of view the only use case we need to target is installing from pypi. If the users install it through their operating system package manager they can ask for an os package variant that includes that package.

Reply:
Also, this is opt-in, so a user will never be surprised by uv on the path or installed when they weren’t trying to use uv. “Try to use local uv the try to use global uv when a user explicitly requests uv” seems quite reasonable.

If you want to just go for one of the two, though, feel free, this is just my opinion (based heavily on experience distributing CMake and ninja packages).

Reply:
Why would non-PyPI distributions include the Python helpers? They aren’t required and require the package to depend on python(s). Homebrew doesn’t tie this to (each) version of Python, for example.

Reply:
And you can’t install a system package manager distribution’s Python helpers into an environment anyway.

Reply:
But then you're saying that this will only be installed into virtual environment, in which case there should be no case when the UV it's not pulled from Pypi and be the latest, and available alongside the build pythons environment. And if that is the case why do we need the ability to specify / use external versions of it? I just don't get it 😅

Reply:
Let's put it this way if the user submits a bug report, what information do we need to understand and replicate the issue? Until now we just require the installed packages into the build project python environment. If all of a sudden we allow external applications even by accident then how do we know what version of UV are they using? Wouldn't be surprising if they're the UV package inside the environment they are running and for whatever reason we are using some external variant available on to the path? The only say I would be okay with specifying external applications / dependencies would be if they would have to be explicitly specified during the invocation inside the CLI.

Reply:
>  it's not pulled from PyPI

The procedure would be, for example on FreeBSD:

```console
$ pkg install uv
$ pipx run build --installer=uv 
```

Ports has uv (https://cgit.freebsd.org/ports/tree/devel/uv), and it _cannot get binary packages from PyPI_; there is no wheel tag for BSD. You cannot use `pipx run build[uv] --installer=uv` on any of the BSDs (at least not without the Rust toolchain, etc). This is the only way to get use it without downloading the Rust toolchain and building from source (assuming it works without dependencies; Ruff has a hidden dependency on GNU Make for a C library that causes issues on minimal docker images).

This is exactly how you build scikit-build-core libraries with cmake and ninja, or meson-python libraries with ninja; you pre-install cmake/ninja of sufficient version from Ports, then you build, and the CLI versions are used instead of requiring them from PyPI - which won't work, since there are no binaries and CMake is required to build them.

I think it's important to view the `uv` PyPI package as simply a distribution of a CLI tool, and not the canonical way to get it. It has a small helper Python file too, but no other distribution includes that - and that's okay, it's just there to make the PyPI package a little friendlier. (And IMO, from cmake and ninja experience, I would have recommended not including a Python helper, I think, it causes lots of arguments about if they should be included in other distributions (which is a clear no, you don't want to force a (specific) Python dependency when there isn't one)).

> Let's put it this way if the user submits a bug report, what information do we need to understand and replicate the issue?

If a user specifies `--installer=uv`, then I think it's perfectly fair to ask them for the output of `pip list` and `uv --version`. It's only an opt-in "advanced" feature.

>  If all of a sudden we allow external applications

Binary builds already use external applications. PyPI doesn't ship compilers. Ruff for example, besides requiring a Rust toolchain, also requires GNU Make (on UNIX, though there's a way to turn of the component that depends on a C lib build). Etc. Installing `build[uv]` requires at least a Rust toolchain.

> The only say I would be okay with specifying external applications / dependencies would be if they would have to be explicitly specified during the invocation inside the CLI.

This is explicitly activated in the CLI. It never runs by default. You are literally passing `--installer=uv`. The only thing we don't do is allow a full path here, it has to be on the PATH if it's not co-installed with build. I think having `--installer=uv` look for `uv` on the path is reasonable.

Reply:
> Wouldn't be surprising if they're the UV package inside the environment they are running and for whatever reason we are using some external variant available on to the path? 

Oh, I do think the uv package should take precedence over the external search. That was a discussion in nox, too, and we've gone with the way that it's been done for years for cmake/ninja, local installs always override external tools.

Reply:
I guess I can accept falling back to the operating system version, but if we do that in my opinion we must print out of warning onto the standard output or error to notify the user that this is happening.

Reply:
Something like this would be perfect, I think:

```
* Creating venv isolated environment...
* Using uv 0.1.16 from /usr/local/bin/uv to install packages...
* Installing packages in isolated environment... (hatch-fancy-pypi-readme, hatch-vcs, hatchling)
```

Whereas if it was from the local environment, it could be simplified:

```
* Creating venv isolated environment...
* Using uv 0.1.16 to install packages...
* Installing packages in isolated environment... (hatch-fancy-pypi-readme, hatch-vcs, hatchling)
```

Would that work?

Reply:
Perfect.

Reply:
Reminder that we have a verbosity flag now that will print out the command invoked including the full uv executable path.  Printing out the version for uv executable from $PATH would require an additional subprocess call and parsing the output of `-V` which is a flimsy affair.  This same consideration also applies to an outer, although the surprise factor is slightly diminished since we’re only looking in the running interpreter’s Python path.

Reply:
<img width="970" alt="Screenshot 2024-03-10 at 11 08 41" src="https://github.com/pypa/build/assets/31134424/4e3b4439-3ecb-4311-b19d-e74b490d6482">

How's this?